### PR TITLE
gd paint from godot 3

### PR DESCRIPTION
**Pull request by Ghofar05**
hello, this is converted project demo of gd paint from godot 3 to godot 4. i've make sure there no bug. thankyou

### CODE REVIEW

File: 2d/gd_paint/paint_control.gd [hunk 1]
```diff
@@ -0,0 +1,253 @@
+extends Control
+
+# A constant for whether or not we're needing to undo a shape.
```
Comment (AThousandShips) [inline 2d/gd_paint/paint_control.gd:3]:
```suggestion

# A constant for whether or not we're needing to undo a shape.
```

File: 2d/gd_paint/tools_panel.gd [hunk 2]
```diff
@@ -0,0 +1,113 @@
+extends Panel
+
+@onready var brush_settings = $BrushSettings
```
Comment (AThousandShips) [inline 2d/gd_paint/tools_panel.gd:3]:
```suggestion

@onready var brush_settings = $BrushSettings
```

File: 2d/gd_paint/tools_panel.gd [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	elif button_name == "mode_rectangle":
+		paint_control.brush_mode = paint_control.BrushModes.RECTANGLE_SHAPE
+		brush_settings.modulate = Color(1, 1, 1, 0.5)
+		tool_name = "Rectangle shape"
+	elif button_name == "mode_circle":
+		paint_control.brush_mode = paint_control.BrushModes.CIRCLE_SHAPE
+		brush_settings.modulate = Color(1, 1, 1, 0.5)
+		tool_name = "Circle shape"
+
+	# If a brush shape button is pressed
```
Comment (AThousandShips) [inline 2d/gd_paint/tools_panel.gd:68]:
```suggestion
	# If a brush shape button is pressed.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

File: 2d/gd_paint/tools_panel.gd [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	elif button_name == "shape_rectangle":
+		paint_control.brush_shape = paint_control.BrushShapes.RECTANGLE
+		shape_name = "Rectangle"
+	elif button_name == "shape_circle":
+		paint_control.brush_shape = paint_control.BrushShapes.CIRCLE
+		shape_name = "Circle";
+
+	# If a opperation button is pressed
```
Comment (AThousandShips) [inline 2d/gd_paint/tools_panel.gd:76]:
```suggestion
	# If a opperation button is pressed.
```

File: 2d/gd_paint/paint_control.gd [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	RECTANGLE_SHAPE,
+}
+
+enum BrushShapes {
+	RECTANGLE,
+	CIRCLE,
+}
+
+# The top-left position of the canvas.
+var TL_node
```
Comment (AThousandShips) [inline 2d/gd_paint/paint_control.gd:24]:
```suggestion
var tl_node
```
`snake_case`

File: 2d/gd_paint/paint_control.gd [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	new_brush.brush_pos = mouse_pos
+	new_brush.brush_shape = brush_shape
+	new_brush.brush_size = brush_size
+	new_brush.brush_color = brush_color
+
+	# If the new bursh is a rectangle shape, we need to calculate the top left corner of the rectangle and the
+	# bottom right corner of the rectangle.
+	if type == BrushModes.RECTANGLE_SHAPE:
+		var TL_pos = Vector2()
+		var BR_pos = Vector2()
```
Comment (AThousandShips) [inline 2d/gd_paint/paint_control.gd:169]:
```suggestion
		var tl_pos = Vector2()
		var br_pos = Vector2()
```

**Comment by Calinou** [comment]
GDPaint was already updated for Godot 4 here, and is fully functional: https://github.com/godotengine/godot-demo-projects/tree/master/gui/gd_paint
I tested it in 4.2.dev5 last week.

**Comment by aaronfranke** [comment]
Calinou is correct, this demo has already been ported and goes in the GUI folder, not the 2D folder, as of https://github.com/godotengine/godot-demo-projects/pull/742.

**Comment by Ghofar05** [comment]
my bad, sorry. i was think this demo have not ported yet because i couldn't find this demo in the godot 4 editor asset library project

**Comment by Calinou** [comment]
> my bad, sorry. i was think this demo have not ported yet because i couldn't find this demo in the godot 4 editor asset library project

See https://github.com/godotengine/godot-demo-projects/issues/920.

**Pull request closed without merging**